### ISSUE TITLE
Handle numbers with customized Decimal & Thousands separators

### ISSUE THREAD

**Issue by geovas01** [fraupflaume/rpivotTable#3]
My problem arises when using the 'TSV Export' Renderer, to paste table contents data to excel, in a computer where the numbers are formatted with "," as the decimal symbol (and "." as the digit grouping symbol).

Ideally, I would suggest adding two extra _parameters_, **Decimal & Thousands separators** when calling `rpivotTable `
(e.g. dec.mark = ",", mark = ".", as in `DT::formatCurrency()` ), if it is possible.

Thank you.

**Comment by mink-lge** [comment]
@fraupflaume If there is a lot of work to be done, in order to implement this for every renderer, I guess it would be perfectly ok to deal only with the TSV one. Also it would be a great help if you added a notification (such as 👀) to let us know that you have seen the request.

**Comment by fraupflaume** [comment]
Alright, I'm still working on it. I wanted you to test what I am thinking is the right direction. I want to incorporate it into the package, but right now I'm testing different approaches. Can you try seeing if you can break this, if it isn't changing all of the values, that sort of thing?

Here's a version...

```
data(gapminder, package = "gapminder")
rpivotTable(gapminder, rows = "continent", aggregatorName = "Average",
            vals = "pop",width = "90%", height = "40%", subtotals = T,
            onRefresh = htmlwidgets::JS("function(el) {
                                          setTimeout(function() {
                                            $('td.pvtVal, td.pvtTotal').each(function() {
                                              var tx = $(this).text();
                                              /*m is the within str match; if , then . else . for ,*/
                                              tx = tx.replace(/[,.]/g, 
                                                            function (m) {return m === ',' ? '.' : ',';}); 
                                              $(this).text(tx);
                                            });
                                          }, 200);}"))
```


Here's another version

```
y = rpivotTable(gapminder, rows = "continent", aggregatorName = "Average",
                vals = "pop",width = "90%", height = "40%", subtotals = T, 
                onRefresh = htmlwidgets::JS("function(el) {
                                              setTimeout(function() {
                                                $('td.pvtVal, td.pvtTotal').each(function() {
                                                  var tx = $(this).text();
                                                        /*ES6 dependent; swap commas and periods */
                                                  tx = tx.replace(/[,.]/g, m => (m === ',' ? '.' : ','));
                                                  $(this).text(tx);
                                                });
                                              }, 200);
                                            }"))

y
```

**Comment by fraupflaume** [comment]
I guess I should also mention that the original method to set number formatting was to choose a locale (or to define your own locale). However, locales are ignored when subtotals are used.

**Comment by mink-lge** [comment]
From what I can make out so far (and before even attemting to test it), you let the control do its job and _afterwards_ you replace "," with "." and vice versa in the cells that are assigned the `pvtVal`, or `pvtTotal` classes. So, graphs will remain unaffected. OK 😐
But what if the cells contain text instead of numbers? 
![image](https://user-images.githubusercontent.com/107258922/173174348-d6638e15-a97d-48ba-b461-285fc3627249.png)

**Comment by fraupflaume** [comment]
That's really interesting (the bar chart problem). All these issues are why I want someone else (or many someones) to run it through the dirt. There are so many ways you can use this table. 
I will look through these known holes and try to find some more patterns I can work with (or create paths to...)
If you find any other holes or potential pitfalls, please share them. 

Thanks!

**Comment by mink-lge** [comment]
> That's really interesting (the bar chart problem). All these issues are why I want someone else (or many someones) to run it through the dirt. There are so many ways you can use this table. I will look through these known holes and try to find some more patterns I can work with (or create paths to...) If you find any other holes or potential pitfalls, please share them.
> 
> Thanks!

I think the safest way is to set a locale. I just don't know how difficult this can be. 
For example, i think it will be difficult to make `onRefresh` corrections in the tsv <textarea>. How could you tell values from attributes and their levels?

**Comment by fraupflaume** [comment]
I'm not planning on leaving it as a function of `onRefresh`. I'm just looking for all of the elements that need to change. 
I would like the user to just tell me what they want the separators to be (or the locale to be). However, I do need to work out the code in the background. I need that before the user interface can be implemented.

You're right. Locale is definitely the best way to change the number formatting. Without subtotal.js, that is how you set the formatting.  However, subtotal.js ignores the preconceived method of locale set by the developers of pivottable.js. I've not had a lot of luck working out the best way to tear down and build up subtotal.js YET. I did want to get something in place while I continue to work through that beast, though. 

As far as `tsv`, that's actually not that difficult. However, it does require a lot of testing (so that I don't miss anything).

For example, this would work `tsv` until it doesn't (breaks or something else I didn't think of yet...).

```
or = htmlwidgets::JS("function(el) {
                      setTimeout(function() {
                        $('td.pvtRendererArea').each(function(){
                          var tx = $(this).last().html();
                          if(tx.indexOf('textarea') != -1) {
                            tx = tx.replace(/[,.]/g, m => (m === ',' ? '.' : ','));
                            $(this).last().html(tx);
                          }
                        });
                        $('td.pvtVal, td.pvtTotal').each(function() {
                          var tx = $(this).text();
                          if (element.children.length == 0) {
                              /*ES6 dependent; swap commas and periods */
                            tx = tx.replace(/[,.]/g, m => (m === ',' ? '.' : ','));
                          } else {
                            tx = tx.lastChild;
                            tx = tx.replace(/[,.]/g, m => (m === ',' ? '.' : ','));
                          }
                          $(this).text(tx);
                        });
                      }, 200);
                    }")
```

**Comment by mink-lge** [comment]
Hmm, not exactly. Notice the unwanted replacements in attribute levels:
![image](https://user-images.githubusercontent.com/107258922/173199047-ff1d1cd0-2a19-4abc-ae2a-af4a74da8b1f.png)

**Comment by fraupflaume** [comment]
How about this... find any holes, let me know! 

```
or = htmlwidgets::JS("function(el) {
                      setTimeout(function() {
                        $('td.pvtRendererArea').each(function(){
                          var tx = $(this).last().html();
                          if(tx.indexOf('textarea') != -1) {
                            console.log(tx);
                            tx = tx.replace(/(?<=\\d)[,.](?=\\d)/g, m => (m === ',' ? '.' : ','));
                            console.log(tx);
                            $(this).last().html(tx);
                          }
                        });
                        $('td.pvtVal, td.pvtTotal').each(function() {
                          var tx = $(this).text();
                          if (element.children.length == 0) {
                              /*ES6 dependent; swap commas and periods */
                            tx = tx.replace(/[,.]/g, m => (m === ',' ? '.' : ','));
                          } else {
                            tx = tx.lastChild;
                            tx = tx.replace(/[,.]/g, m => (m === ',' ? '.' : ','));
                          }
                          $(this).text(tx);
                        });
                      }, 200);
                    }")
```

**Comment by mink-lge** [comment]
At a first glance, I believe there will be a problem in a case of e.g. a categorical variable that contained intervals 
`... ,"(40,50]", "(50,60]", "(60,70]", ...` 
I will update if I find more after testing.

**Comment by fraupflaume** [comment]
This is all very helpful! Thank you! I'm not sure how I would go about addressing this. However, knowing it's an issue is half the battle.
I came across something interesting while working through the subtotal.js. It was interesting enough for me to set aside all of what I was working on. When I run `subtotals = T` and add a `locale`, the locale is not ignored. I don't know if this was working initially or not. I know that there are comments in the package that say it does not work this way. I do want to point out that I it does *not* help when it comes to the TSV renderer. However, if that is the only script I have to address, that still makes things a LOT easier.
Can you tell me if the locale is ignored or recognized for you when `subtotals = T`? (Assuming you're using my fork of this package.)
Thanks again!

**Comment by mink-lge** [comment]
It seems that the main purpose of the locale option, is to provide the aggregators in different languages (AND the appropriate number formatting).
Unfortunately "gr" is not included in the valid locales.
If I give e.g. `.., aggregatorName = "Gennemsnit", locale = "da", ...` I get the desired formatting, but it doesn't help much :)
Αs you rightly said, locale has no effect on the TSV renderer. 

Thank you!

**Comment by fraupflaume** [comment]
I can add Greek number formatting, along with renderers and aggregators. That's not all that difficult actually! It may sound a bit better if you were to tell me the colloquial terms (you know, instead of Google Translate). 

As far as the column names that are numbers, I would have expected them to be ignored. The idea that these are names would lead me to believe they shouldn't be manipulated, either. 

So that leaves me to work out a Greek locale, manipulating the TSV renderer, and manipulating the unique values aggregator.

Just to make sure I'm on the right track, a numbered list in most of the world would be separated by semi-colon's because of the comma decimal separator, right? As in 21,4; 21,5; 22,8; etc.?

I think I may be able to get the other renderer names to be shown in the language selected, as well. (They are all language-specific when subtotals is not selected.)

**Comment by mink-lge** [comment]
All that sound like a lot of work to me. Wouldn't be easier to leave everything, except number formatting, as it is for the "en" locale? _Personally I find easier to understand the terms in English than in Greek :)_
I agree with the semi-colon separator for lists.

**Comment by fraupflaume** [comment]
Alright, if you re-install this package repo, you can select `gr` as your locale. It's the Greek locale, but the terms are all in English (the number formatting is as *most* of the world, not American). TSV will respect the formatting you assign by locale. I did not change the number formatting for unique values based on your last comment (I did change the item sep from `,` to `;`, though). I can still change the number formatting for this aggregator—just let me know. 

(Well, there is a bit of Greek, but you have to either cheat (look at the JS) or break something to see it :)

Other than a Greek scavenger hunt, I would appreciate it if you took a look a the changes and see if there is something I didn't account for. What did I miss?

**Comment by mink-lge** [comment]
> Alright, if you re-install this package repo, you can select `gr` as your locale. It's the Greek locale, but the terms are all in English (the number formatting is as _most_ of the world, not American). TSV will respect the formatting you assign by locale. I did not change the number formatting for unique values based on your last comment (I did change the item sep from `,` to `;`, though). I can still change the number formatting for this aggregator—just let me know.
> 
> (Well, there is a bit of Greek, but you have to either cheat (look at the JS) or break something to see it :)
> 
> Other than a Greek scavenger hunt, I would appreciate it if you took a look a the changes and see if there is something I didn't account for. What did I miss?

I just don't know how to thank you @fraupflaume 🤗
(I saw some error messages in greek on pivot.gr.min.js)
I am going to try it now, and I'll let you know if something is wrong (I still believe `,` is better than `;` for lists, though)

**Comment by mink-lge** [comment]
Well, I have already found something. The "." (thousands separator) is missing. (a blank space appears instead). This causes problems when numbers are transferred to excel (they are not correctly recognized). I guess it must be a simple fix.
![image](https://user-images.githubusercontent.com/107258922/174001144-9d84df3e-23ce-479c-a2de-83d562c6775e.png)

**Comment by fraupflaume** [comment]
Alright, I had to go about modifying the tsv renderer output a different way. However, I think I've covered all of the requests you have. If you wouldn't mind—please install the latest and let me know!

**Comment by mink-lge** [comment]
Apart from this minor string error, everything else looks perfect! 
![image](https://user-images.githubusercontent.com/107258922/174226029-24d074e4-3eeb-4b4b-af96-674e81d0e015.png)
Please do this last correction and close this issue.
Thank you very much, indeed :)

**Comment by geovas01** [comment]
Thank you all, very very much!

**Comment by fraupflaume** [comment]
Ya, it's all Greek to me, too. LOL!! That really shouldn't be as funny as it is!!

So weird with the error on that same error message. I remembered 'fixing' it. I copied and pasted it again, and noticed the same thing happened. I mean sheesh! I had to delete the whole line. I guess it's because of the left-to-right vs right-to-left deal. (Which I didn't know about Greek until now.)

All fixed! Glad I could help you all out!
